research-prototype/Programs/Source/ipae2e.mpc

```diff
@@ -0,0 +1,154 @@
+
+for i in range(numrows-1):
+    cmd = f"helperbit_and_istrigger[{i}] = helperbit[{i}] * _list[{i}][1]"
+    exec(cmd, {}, local_params)
+    cmd = f"helperbit_xor_istrigger[{i}] = (helperbit[{i}] + _list[{i}][1]) % 2"
+    exec(cmd, {}, local_params)
+
+if verbose:
+    print_ln('%s', helperbit.reveal())
+    print_ln('%s', helperbit_and_istrigger.reveal())
+    print_ln('%s', helperbit_xor_istrigger.reveal())
+
+
+stopbits = [Array(numrows, sint) for _ in range(numrows_power)]
+local_params['stopbits'] = stopbits
+
+for stopbit in stopbits:
+    local_params['stopbit'] = stopbit
+    for i in range(numrows):
+        cmd = f"stopbit[{i}] = 1"
+        exec(cmd, {}, local_params)
+
+credits = [Array(numrows, sint) for _ in range(numrows_power)]
+local_params['credits'] = credits
+
+for i, (stopbit, credit) in enumerate(zip(stopbits, credits)):
+    stepsize = 2 ** i
+    local_params['stopbit'] = stopbit
+    local_params['credit'] = credit
+    local_params['stepsize'] = stepsize
+
+    for current in range(numrows - stepsize):
+        successor = current + stepsize
+        # another optimization could be to multiply one time all the
+        # helperbit * istrigger bits since these are fixed.
+        if i == 0:
+            cmd = f"flag = helperbit_and_istrigger[{successor}];"
+            # credit is initialized to value, _list[i][2] = value[i]
+            cmd += f"credit[{current}] = _list[{current}][2] + flag * _list[{successor}][2];"
+            # * stopbit[successor]  stopbit[successor] here would always be 1 so can drop
+            cmd += f"stopbit[{current}] = flag"
+        else:
+            cmd = f"flag = stopbits[{i-1}][{current}] * helperbit_and_istrigger[{successor}];"
+            cmd += f"credit[{current}] = "
+            cmd += f"credits[{i-1}][{current}] + flag * credits[{i-1}][{successor}];"
+            cmd += f"stopbit[{current}] = flag * stopbits[{i-1}][{successor}]"
+
+        exec(cmd, {}, local_params)
+
+    if verbose:
+        print_ln('iteration: %s', i)
+        print_ln('credit %s', credit.reveal())
+        print_ln('stopbit %s', stopbit.reveal())
+
+# I can't sum all of credit, I need to sum only those that correspond to last touches
+# helperbit_xor_istrigger * credits[-1] will give this.
+creditfinal = Array(numrows, sint)
+local_params['creditfinal'] = creditfinal
+
+for i in range(numrows):
+    cmd = f"creditfinal[{i}] = helperbit_xor_istrigger[{i}] * credits[-1][{i}]"
+    exec(cmd, {}, local_params)
+
+# SUM BREAKDOWNS
+# for now we have hard coded 4 breakdown keys
+breakdowns = 4
+
+bksums = Array(4, sint)
+local_params['bksums'] = bksums
+
+for bk in range(len(bksums)):
+    cmd = f"bksums[{bk}] = sint(0)"
+    exec(cmd, {}, local_params)
+    for i in range(numrows):
+        cmd = (
+            f"bksums[{bk}] = bksums[{bk}] + "
+            f"(_list[{i}][3] == {bk}).if_else(creditfinal[{i}], 0)"
+        )
+        exec(cmd, {}, local_params)
```

Comment:
@bmcase I'm 99% sure this does the same thing as before (but doesn't do the "tree" approach. Given that we have to do the `if_else` on every row anyways, I'm not sure it really added any efficiency, and this is definitely easier to follow. Let me know if that makes sense.